### PR TITLE
feat: add flow to initialize CNS from CNI

### PR DESCRIPTION
**Pull request by rbtr**
**Reason for Change**:

- [x] Updates the CNS pod statemap to be able to use the new interface ID being passed from CNI instead of keying on Kubernetes namespace/name
- [x] Adds a flow where CNS can rebuild its internal state from CNI instead of from the Kubernetes API
- [x] Adds a config flag and a CNI version check, making the above two changes explicitly opt-in

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests

**Notes**:

### CODE REVIEW

File: cni/api/api.go [hunk 1]
```diff
@@ -9,23 +9,19 @@ import (
...
-}
-
-type AzureCNIState struct {
+type State struct {
 	ContainerInterfaces map[string]PodNetworkInterfaceInfo
 }
 
-func (a *AzureCNIState) PrintResult() error {
-	b, err := json.MarshalIndent(a, "", "    ")
+func (s *State) WriteToStdout() error {
```
Comment (neaggarwMS) [inline cni/api/api.go:23]: There are CNI changes in this PR. Does that mean you will cut CNI release before CNS?

File: cns/NetworkContainerContract.go [hunk 2]
```diff
@@ -107,16 +107,130 @@ type ConfigureContainerNetworkingRequest struct {
 	NetworkContainerid string
 }
 
-// KubernetesPodInfo is an OrchestratorContext that holds PodName and PodNamespace.
-type KubernetesPodInfo struct {
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract.go:111]: CNS is used in other orchestrators as well like ACI, So not sure if we should rename the struct name.


For some scenarios, Pod Key is still name + namespace

File: cns/service/main.go [hunk 3]
```diff
@@ -405,6 +410,25 @@ func main() {
 	configuration.SetCNSConfigDefaults(&cnsconfig)
 	logger.Printf("[Azure CNS] Read config :%+v", cnsconfig)
 
+	// We might be configured to reinitialize state from the CNI instead of the apiserver.
+	// If so, we should check that the the CNI is new enough to support the state commands,
+	// otherwise we fall back to the existing behavior.
+	if cnsconfig.InitializeFromCNI {
+		isGoodVer, err := cni.IsDumpStateVer()
```
Comment (neaggarwMS) [inline cns/service/main.go:417]: This is only checked if ChannelMode is cns.CRD.

Comment (neaggarwMS) [reply cns/service/main.go:417]: For other configurations (like for ACI), reconcilation happens between DNC <> CNS

File: cns/service/main.go [hunk 4]
```diff
@@ -506,9 +530,7 @@ func main() {
 	// Initialze state in if CNS is running in CRD mode
 	// State must be initialized before we start HTTPRestService
 	if config.ChannelMode == cns.CRD {
-		requestControllerStopChannel := make(chan struct{})
-		defer close(requestControllerStopChannel)
-		err = InitializeCRDState(httpRestService, cnsconfig, requestControllerStopChannel)
+		err = InitializeCRDState(ctx, httpRestService, cnsconfig)
```
Comment (neaggarwMS) [inline cns/service/main.go:533]: These are great fixes but making this PR really bulky. Is it possible to move this to a new PR? If not, then may be in future we should take care of it.

File: cns/service/main.go [hunk 5]
```diff
@@ -855,12 +880,10 @@ func InitializeCRDState(httpRestService cns.HTTPService, cnsconfig configuration
...
 			case <-tickerChannel:
 				httpRestServiceImplementation.SyncHostNCVersion(rootCxt, cnsconfig.ChannelMode, cnsconfig.SyncHostNCTimeoutMs)
-            case <-rootCxt.Done():
-                return
+			case <-rootCxt.Done():
+				return
 			}
 		}
-
-		logger.Printf("[Azure CNS] Exiting SyncHostNCVersion")
```
Comment (neaggarwMS) [inline cns/service/main.go:863]: Why did we remove this log line?

File: cns/service/main.go [hunk 6]
```diff
@@ -405,6 +410,25 @@ func main() {
...
+			logger.Errorf("error checking CNI ver: %v", err)
+		}
+
+		// override the prior config flag with the result of the ver check.
+		cnsconfig.InitializeFromCNI = isGoodVer
+
+		if cnsconfig.InitializeFromCNI {
+			// Set the PodInfoVersion by initialization type, so that the
+			// PodInfo maps use the correct key schema
+			cns.GlobalPodInfoVersion = cns.InterfaceIDPodInfoVersion
```
Comment (neaggarwMS) [inline cns/service/main.go:428]: NIT: Can we rename it to GlobalPodInfoKeyType?   PodInfoVersion sounds more like Pod version

Comment (neaggarwMS) [inline cns/service/main.go:428]: Can you add a log line here dumping what PodKey we chose during init

File: cns/restserver/util.go [hunk 7]
```diff
@@ -164,8 +164,7 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
 		case cns.AzureFirstParty:
 			fallthrough
 		case cns.WebApps: // todo: Is WebApps an OrchastratorType or ContainerType?
-			var podInfo cns.KubernetesPodInfo
```
Comment (neaggarwMS) [inline cns/restserver/util.go:167]: Did you also validate ACI and WebApps scenario to ensure there is no breaking change there?

File: cns/restserver/util.go [hunk 8]
```diff
@@ -177,7 +176,7 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
 				service.state.ContainerIDByOrchestratorContext = make(map[string]string)
 			}
 
-			service.state.ContainerIDByOrchestratorContext[podInfo.PodName+podInfo.PodNamespace] = req.NetworkContainerid
```
Comment (neaggarwMS) [inline cns/restserver/util.go:180]: >[podInfo.PodName+podInfo.PodNamespace [](http://example.com/codeflow?start=49&length=37)

This will fail the ACI scenario: previously Key is PodName+PodNamespae (Pod1podns).....and now the Key is PodName:PodNameSpace......

Please note this map is stored in the statefile, so ACI update will fail

File: cns/restserver/util.go [hunk 9]
```diff
@@ -359,8 +357,7 @@ func (service *HTTPRestService) getNetworkContainerResponse(req cns.GetNetworkCo
 
 		logger.Printf("pod info %+v", podInfo)
 
-		context := podInfo.PodName + podInfo.PodNamespace
-		containerID, exists = service.state.ContainerIDByOrchestratorContext[context]
```
Comment (neaggarwMS) [inline cns/restserver/util.go:363]: >containerID, exists = service.state.ContainerIDByOrchestratorContext[context] [](http://example.com/codeflow?start=2&length=77)

This is a breaking change, can change the Key

File: cns/restserver/util.go [hunk 10]
```diff
@@ -671,22 +667,20 @@ func (service *HTTPRestService) SendNCSnapShotPeriodically(ncSnapshotIntervalInM
...
 	}
 
 	if ipConfigRequest.OrchestratorContext == nil {
-		return podInfo, EmptyOrchestratorContext, fmt.Sprintf("OrchastratorContext is not set in the req: %+v", ipConfigRequest)
+		return nil, EmptyOrchestratorContext, fmt.Sprintf("OrchastratorContext is not set in the req: %+v", ipConfigRequest)
 	}
 
 	// retrieve podinfo  from orchestrator context
-	if err := json.Unmarshal(ipConfigRequest.OrchestratorContext, &podInfo); err != nil {
+	podInfo, err := cns.NewPodInfoFromIPConfigRequest(ipConfigRequest)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:680]: If the PodInfoVersion is set to InterfaceId, then can you check if PodInfo InterfaceId is not empty? May be do it inside Unmarshal wrapper

File: cns/requestcontroller/kubecontroller/crdrequestcontroller.go [hunk 11]
```diff
@@ -239,40 +241,52 @@ func (crdRC *crdRequestController) initCNS() error {
...
-		for _, pod = range pods.Items {
-			//Only add pods that aren't on the host network
-			if !pod.Spec.HostNetwork {
-				podInfo = cns.KubernetesPodInfo{
-					PodName:      pod.Name,
-					PodNamespace: pod.Namespace,
-				}
-				podInfoByIP[pod.Status.PodIP] = podInfo
-			}
+	if crdRC.cfg.InitializeFromCNI {
```
Comment (neaggarwMS) [inline cns/requestcontroller/kubecontroller/crdrequestcontroller.go:252]: I dont think this alone is definining how to reconcile. Rather you need to check GlobalPodInfoVersion

File: cns/cni/initialize.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"github.com/Azure/azure-container-networking/cni/api"
+	"github.com/Azure/azure-container-networking/cni/client"
+	"github.com/Azure/azure-container-networking/cns"
+	"k8s.io/utils/exec"
+)
+
+// CNICNSInitializer returns an implementation of cns.PodInfoByIPProvider
+// that execs out to the CNI and uses the response to build the PodInfo map.
+func CNICNSInitializer() (cns.PodInfoByIPProvider, error) {
```
Comment (neaggarwMS) [inline cns/cni/initialize.go:12]: Can we change the folder name from cns/cni to be something specific (for lack of better word, cnireconciler?

Comment (neaggarwMS) [inline cns/cni/initialize.go:12]: CNIPodInfoReconciler?

File: cns/cni/initialize.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+	}), nil
+}
+
+// cniStateToPodInfoByIP converts an AzureCNIState dumped from a CNI exec
+// into a PodInfo map, using the first endpoint IP as the key in the map.
+func cniStateToPodInfoByIP(state *api.State) map[string]cns.PodInfo {
+	podInfoByIP := map[string]cns.PodInfo{}
+	for _, endpoint := range state.ContainerInterfaces {
+		podInfoByIP[endpoint.IPAddresses[0].IP.String()] = cns.NewPodInfo(
+			"",
```
Comment (neaggarwMS) [inline cns/cni/initialize.go:29]: >"", [](http://example.com/codeflow?start=3&length=3)

Why are we not populating PodInfraContainerID?

File: cns/cni/initialize.go [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/Azure/azure-container-networking/cni/client"
+	"github.com/Azure/azure-container-networking/cns"
+	"k8s.io/utils/exec"
+)
+
+// CNICNSInitializer returns an implementation of cns.PodInfoByIPProvider
+// that execs out to the CNI and uses the response to build the PodInfo map.
+func CNICNSInitializer() (cns.PodInfoByIPProvider, error) {
+	cli := client.New(exec.New())
+	state, err := cli.GetEndpointState()
```
Comment (neaggarwMS) [inline cns/cni/initialize.go:14]: @matmerr, does CNI acquires the lock in GetEndpointState?

File: cns/service/main.go (hunk 5 above)
Comment (rbtr) [reply cns/service/main.go:863]: it is unreachable

File: cni/api/api.go (hunk 1 above)
Comment (rbtr) [reply cni/api/api.go:23]: the CNI changes are superficial changes to the go client and do not need a CNI release or new binary
/cc @matmerr

Comment (matmerr) [reply cni/api/api.go:23]: yeah I don't really have any opinions here, but potentially add a comment in addition to the rename mentioning `PrintResult` or `Print*`, as [CNI writes results to stdout via `PrintResult(result, version)` or more directly `result.Print()`](https://github.com/containernetworking/cni/blob/master/pkg/types/types.go#L103), and given the nonstandard cmd we're adding I'd like to have some trail to regex with when tracking down where CNI is returning a response

**Comment by neaggarwMS** [comment]
> package network

What is the change in this file? Can you revert it if there are no changes

---
Refers to: cni/network/network_test.go:1 in 44e8e20. [](commit_id = 44e8e204acacfd21063a7ffa1a11d50d29d44d58, deletion_comment = False)

File: cni/network/invoker_cns.go [hunk 15]
```diff
@@ -179,14 +180,14 @@ func setHostOptions(nwCfg *cni.NetworkConfig, hostSubnetPrefix *net.IPNet, ncSub
...
+	podInfo := cns.NewPodInfo(args.ContainerID, endpointId, invoker.podName, invoker.podNamespace)
 
 	orchestratorContext, err := json.Marshal(podInfo)
 	if err != nil {
 		return err
 	}
 
-	endpointId := GetEndpointID(args)
 	req := cns.IPConfigRequest{
 		OrchestratorContext: orchestratorContext,
```
Comment (neaggarwMS) [inline cni/network/invoker_cns.go:192]: The orchestrator context was initillay podname + Podns, now it is the whole endpoint id, containerId and podname and podns....


I think in CNS we still treat orchestrator context to be podname+podns only

File: cni/client/client.go [hunk 16]
```diff
@@ -14,21 +14,25 @@ import (
 )
 
 const (
-	azureVnetBinName      = "./azure-vnet"
-	azureVnetBinDirectory = "/opt/cni/bin"
+	azureVnetExecutable = "/opt/cni/bin/azure-vnet"
```
Comment (neaggarwMS) [inline cni/client/client.go:17]: Does this client gets packaged with CNI too?

File: cni/network/invoker_cns.go [hunk 17]
```diff
@@ -51,14 +51,15 @@ func NewCNSInvoker(podName, namespace string) (*CNSIPAMInvoker, error) {
...
-	podInfo := cns.KubernetesPodInfo{PodName: invoker.podName, PodNamespace: invoker.podNamespace}
+	podInfo := cns.NewPodInfo(args.ContainerID, endpointId, invoker.podName, invoker.podNamespace)
 	orchestratorContext, err := json.Marshal(podInfo)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	endpointId := GetEndpointID(args)
 	ipconfig := cns.IPConfigRequest{
 		OrchestratorContext: orchestratorContext,
```
Comment (neaggarwMS) [inline cni/network/invoker_cns.go:64]: Similarly here, OrchestratorContext info is changed here. It is meant to be podName + PodNS

File: cni/network/multitenancy.go [hunk 18]
```diff
@@ -64,18 +64,14 @@ func getContainerNetworkConfiguration(
...
-	ifName string) (*cniTypesCurr.Result, *cns.GetNetworkContainerResponse, net.IPNet, error) {
+func getContainerNetworkConfigurationInternal(address string, namespace string, podName string, ifName string) (*cniTypesCurr.Result, *cns.GetNetworkContainerResponse, net.IPNet, error) {
 	cnsClient, err := cnsclient.GetCnsClient()
 	if err != nil {
 		log.Printf("Failed to get CNS client. Error: %v", err)
 		return nil, nil, net.IPNet{}, err
 	}
 
-	podInfo := cns.KubernetesPodInfo{PodName: podName, PodNamespace: namespace}
+	podInfo := cns.NewPodInfo("", "", podName, namespace)
```
Comment (neaggarwMS) [inline cni/network/multitenancy.go:74]: I dont think we can change this, existing partners CNI's are not updated to latest CNI version, so this will break with just cns update

File: cns/restserver/util.go [hunk 19]
```diff
@@ -165,8 +165,7 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
 		case cns.AzureFirstParty:
 			fallthrough
 		case cns.WebApps: // todo: Is WebApps an OrchastratorType or ContainerType?
-			var podInfo cns.KubernetesPodInfo
```
Comment (neaggarwMS) [inline cns/restserver/util.go:168]: Again, I dont think we can change this. Unmarshal will fail as OrchestratorContext is only pod name + podns

File: cns/NetworkContainerContract.go (hunk 2 above)
Comment (neaggarwMS) [reply cns/NetworkContainerContract.go:111]: I'm worried for using the new struct in other scenarios. can we create a separate struct for AKS scenario, then later converge to one.

File: cni/client/client.go (hunk 16 above)
Comment (rbtr) [reply cni/client/client.go:17]: no, it's only imported by cns

File: cns/cni/initialize.go (hunk 14 above)
Comment (rbtr) [reply cns/cni/initialize.go]: it looks like it does since CNI is just invoked normally and it always takes the lock

File: cns/NetworkContainerContract_test.go [hunk 20]
```diff
@@ -0,0 +1,44 @@
+package cns
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUnmarshalPodInfo(t *testing.T) {
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract_test.go:10]: Can you also add test for NewPodInfoFromIPConfigRequest?

File: cns/cnireconciler/podinfoprovider.go [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+	"k8s.io/utils/exec"
+)
+
+// NewCNIPodInfoProvider returns an implementation of cns.PodInfoByIPProvider
+// that execs out to the CNI and uses the response to build the PodInfo map.
+func NewCNIPodInfoProvider() (cns.PodInfoByIPProvider, error) {
+	cli := client.New(exec.New())
+	state, err := cli.GetEndpointState()
+	if err != nil {
+		return nil, err
```
Comment (neaggarwMS) [inline cns/cnireconciler/podinfoprovider.go:16]: Can we return fmt.ErrorF and append the error with more info say "Failed to invoke CNI.GetEndpointsate(), error: %s", err?

File: cns/cnireconciler/podinfoprovider.go [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"github.com/Azure/azure-container-networking/cni/api"
+	"github.com/Azure/azure-container-networking/cni/client"
+	"github.com/Azure/azure-container-networking/cns"
+	"k8s.io/utils/exec"
+)
+
+// NewCNIPodInfoProvider returns an implementation of cns.PodInfoByIPProvider
+// that execs out to the CNI and uses the response to build the PodInfo map.
+func NewCNIPodInfoProvider() (cns.PodInfoByIPProvider, error) {
```
Comment (neaggarwMS) [inline cns/cnireconciler/podinfoprovider.go:12]: Also is it possible to add UT for this?

File: cns/cnsclient/cnsclient_test.go [hunk 23]
```diff
@@ -225,7 +225,7 @@ func TestCNSClientRequestAndRelease(t *testing.T) {
 
 	addTestStateToRestServer(t, secondaryIps)
 
-	podInfo := cns.KubernetesPodInfo{PodName: podName, PodNamespace: podNamespace}
```
Comment (neaggarwMS) [inline cns/cnsclient/cnsclient_test.go:228]: We need to revert this and pass the way CNI passes it, this will KubernetesPodInfo only, and can we pass PodInterfaceIf and InfraContainerId in the ipConfigRequest.

It will be ideal to run this test for both old and new behaviour, but if we cant then at least move them to new as in production we are sure it will be new version

File: cns/configuration/cns_config.json [hunk 24]
```diff
@@ -1,23 +1,24 @@
 {
     "TelemetrySettings": {
```
Comment (neaggarwMS) [inline cns/configuration/cns_config.json:2]: Why so many changes? Is it not rebased to master?

File: cns/restserver/api_test.go [hunk 25]
```diff
@@ -706,7 +707,7 @@ func createOrUpdateNetworkContainerWithParams(t *testing.T, params createOrUpdat
 	ipSubnet.IPAddress = params.ncIP
 	ipSubnet.PrefixLength = 24
 	ipConfig.IPSubnet = ipSubnet
-	podInfo := cns.KubernetesPodInfo{PodName: "testpod", PodNamespace: "testpodnamespace"}
```
Comment (neaggarwMS) [inline cns/restserver/api_test.go:709]: These are multitenant scenarios, we dont need these changes as marshalling KubernetesPodInfo should work too, can you revert these changes?

File: cns/restserver/api_test.go [hunk 26]
```diff
@@ -774,9 +775,7 @@ func deleteNetworkContainerWithParams(t *testing.T, params createOrUpdateNetwork
 func getNetworkContainerByContext(t *testing.T, params createOrUpdateNetworkContainerParams) error {
 	var body bytes.Buffer
 	var resp cns.GetNetworkContainerResponse
-	podInfo := cns.KubernetesPodInfo{
```
Comment (neaggarwMS) [inline cns/restserver/api_test.go:777]: Revert this change too.

File: cns/restserver/api_test.go [hunk 27]
```diff
@@ -803,9 +802,7 @@ func getNetworkContainerByContext(t *testing.T, params createOrUpdateNetworkCont
 func getNonExistNetworkContainerByContext(t *testing.T, params createOrUpdateNetworkContainerParams) error {
 	var body bytes.Buffer
 	var resp cns.GetNetworkContainerResponse
-	podInfo := cns.KubernetesPodInfo{
```
Comment (neaggarwMS) [inline cns/restserver/api_test.go:806]: Revert

File: cns/restserver/api_test.go [hunk 28]
```diff
@@ -832,9 +829,7 @@ func getNonExistNetworkContainerByContext(t *testing.T, params createOrUpdateNet
 func getNetworkContainerByContextExpectedError(t *testing.T, params createOrUpdateNetworkContainerParams) error {
 	var body bytes.Buffer
 	var resp cns.GetNetworkContainerResponse
-	podInfo := cns.KubernetesPodInfo{
-		PodName:      params.podName,
```
Comment (neaggarwMS) [inline cns/restserver/api_test.go:836]: Here too

**Comment by neaggarwMS** [comment]
> func TestReconcileNCWithExistingState(t *testing.T) {

Can we update this test to new struct (pass InterfaceId/ContainerId and validate that as well)

---
Refers to: cns/restserver/internalapi_test.go:220 in 6f27e8a. [](commit_id = 6f27e8af4e2ef1755f5b50c5e7f99574ae0a76e1, deletion_comment = False)

**Comment by neaggarwMS** [comment]
> 		logger.Errorf("[releaseIPConfig] SetIPConfigAsAvailable failed to release, no allocation found for pod [%+v]", podInfo)

It is okay to get idempotent release calls, can you update this message for "Ignore release as no allocations found for this pod"

---
Refers to: cns/restserver/ipam.go:397 in 6f27e8a. [](commit_id = 6f27e8af4e2ef1755f5b50c5e7f99574ae0a76e1, deletion_comment = False)

File: cns/restserver/ipam.go [hunk 29]
```diff
@@ -55,46 +53,33 @@ func (service *HTTPRestService) requestIPConfigHandler(w http.ResponseWriter, r
...
 
-	podInfo, statusCode, returnMessage := service.validateIpConfigRequest(req)
+	var podInfo cns.PodInfo
+	podInfo, resp.ReturnCode, resp.Message = service.validateIpConfigRequest(req)
 
 	if err = service.releaseIPConfig(podInfo); err != nil {
-		statusCode = NotFound
-		returnMessage = err.Error()
-		logger.Errorf("releaseIPConfigHandler releaseIPConfig failed because %v, release IP config info %s", returnMessage, req)
+		resp.ReturnCode = NotFound
```
Comment (neaggarwMS) [inline cns/restserver/ipam.go:78]: May be we should change it to UnexpectedError, NotFound is accepted behaviour

**Comment by neaggarwMS** [comment]
> var (

We should either run all these test with old and new contract or move them to new (where you also pass interfaceId and infracontainer in IPConfigRequest"

---
Refers to: cns/restserver/ipam_test.go:16 in 6f27e8a. [](commit_id = 6f27e8af4e2ef1755f5b50c5e7f99574ae0a76e1, deletion_comment = False)

File: cns/restserver/util.go [hunk 30]
```diff
@@ -403,7 +395,7 @@ func (service *HTTPRestService) getNetworkContainerResponse(req cns.GetNetworkCo
 				return getNetworkContainerResponse
 			}
 
-			containerID = service.state.ContainerIDByOrchestratorContext[context]
+			containerID = service.state.ContainerIDByOrchestratorContext[podInfo.Key()]
```
Comment (neaggarwMS) [inline cns/restserver/util.go:398]: >podInfo.Key() [](http://example.com/codeflow?start=64&length=13)

This needs to be changed to Name+Namesapce again

File: cns/cnireconciler/version.go [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+// state and returns the result of that test or an error. Will always
+// return false when there is an error.
+func IsDumpStateVer() (bool, error) {
+	needVer, err := semver.NewVersion(cniDumpStateVer)
+	if err != nil {
+		return false, err
+	}
+	cnicli := client.New(exec.New())
+	if ver, err := cnicli.GetVersion(); err != nil {
+		return false, err
```
Comment (neaggarwMS) [inline cns/cnireconciler/version.go:22]: Can we fmt.ErrorF and append the error message

**Review by neaggarwMS — CHANGES_REQUESTED** [review]
:clock1:

File: cns/service/main.go [hunk 32]
```diff
@@ -426,6 +428,31 @@ func main() {
 	configuration.SetCNSConfigDefaults(&cnsconfig)
 	logger.Printf("[Azure CNS] Read config :%+v", cnsconfig)
 
+	// We might be configured to reinitialize state from the CNI instead of the apiserver.
+	// If so, we should check that the the CNI is new enough to support the state commands,
+	// otherwise we fall back to the existing behavior.
+	if cnsconfig.InitializeFromCNI {
```
Comment (neaggarwMS) [inline cns/service/main.go:434]: Can you move this block only if cns mode is ViaCRD

**Review by neaggarwMS — CHANGES_REQUESTED** [review]
:clock1:

File: cns/configuration/cns_config.json (hunk 24 above)
Comment (rbtr) [reply cns/configuration/cns_config.json:2]: only change is putting the fields in lexical order

File: cns/cnireconciler/podinfoprovider_test.go [hunk 33]
```diff
@@ -0,0 +1,52 @@
...
+func newCNIStateFakeExec(stdout string) exec.Interface {
+	calls := []testutils.TestCmd{
+		{Cmd: []string{"./azure-vnet"}, Stdout: stdout},
+	}
+
+	fake, _ := testutils.GetFakeExecWithScripts(calls)
+	return fake
+}
+
+func TestNewCNIPodInfoProvider(t *testing.T) {
```
Comment (neaggarwMS) [inline cns/cnireconciler/podinfoprovider_test.go:21]: Thanks for adding the tests :)

File: cns/restserver/internalapi_test.go [hunk 34]
```diff
@@ -235,15 +234,42 @@ func TestReconcileNCWithExistingState(t *testing.T) {
...
+		secIpConfig := newSecondaryIPConfig(ipaddress, -1)
+		ipId := uuid.New()
+		secondaryIPConfigs[ipId.String()] = secIpConfig
+		startingIndex++
+	}
+	req := generateNetworkContainerRequest(secondaryIPConfigs, "reconcileNc1", "-1")
+
+	expectedAllocatedPods := map[string]cns.PodInfo{
+		"10.0.0.6": cns.NewPodInfo("reconc-eth0", "abcdef", "reconcilePod1", "PodNS1"),
+		"10.0.0.7": cns.NewPodInfo("reconc-eth0", "abcxyz", "reconcilePod2", "PodNS1"),
```
Comment (neaggarwMS) [inline cns/restserver/internalapi_test.go:272]: Two ips are allocated to same interfaceID, unsupported scenario

Comment (neaggarwMS) [reply cns/restserver/internalapi_test.go:272]: Can you change the interfaceId for the second ip

Comment (neaggarwMS) [reply cns/restserver/internalapi_test.go:272]: Is this test passing?

Comment (rbtr) [reply cns/restserver/internalapi_test.go:272]: yes, it's passing

**Review by neaggarwMS — DISMISSED** [review]

**Review by neaggarwMS — APPROVED** [review]
:shipit:
